unitary/alpha/quokka_sampler.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2024 The Unitary Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Simulation using a "Quokka" device."""
+
+from typing import Any, Callable, Dict, Optional, Sequence
+
+import cirq
+import numpy as np
+import json
+
+_REQUEST_ENDPOINT = "http://{}.quokkacomputing.com/qsim/qasm"
+_DEFAULT_QUOKKA_NAME = "quokka1"
+
+JSON_TYPE = Dict[str, Any]
+_RESULT_KEY = "result"
+_ERROR_CODE_KEY = "error_code"
+_RESULT_KEY = "result"
+_SCRIPT_KEY = "script"
+_REPETITION_KEY = "count"
+
+
+class QuokkaSampler(cirq.Sampler):
+    """Sampler for querying a Quokka quantum simulation device.
+
+    See https://www.quokkacomputing.com/ for more information.a
+
+    Args:
+        name: name of your quokka device
+        endpoint: HTTP url endpoint to post queries to.
+        post_function: used only for testing to override default
+            behavior to connect to internet URLs.
+    """
+
+    def __init__(
+        self,
+        name: str = _DEFAULT_QUOKKA_NAME,
+        endpoint: Optional[str] = None,
+        post_function: Optional[Callable[[JSON_TYPE], JSON_TYPE]] = None,
+    ):
+        self.quokka_name = name
+        self.endpoint = endpoint
+        self.post_function = post_function
+
+        if self.endpoint is None:
+            self.endpoint = _REQUEST_ENDPOINT.format(self.quokka_name)
+        if self.post_function is None:
+            self.post_function = self._post
```

Comment:
this, combined with the _post method below, is a bit weird. A cleaner implementation would be to extract _post method into the module level function, and then pass it as a default value of post_function argument. 

Maybe something like that:

```py
class Quokka(object):
  def __init__(self, name=_DEFAULT_QUOKKA_NAME):
    self._end_point = _REQUEST_ENDPOINT.format(name)
  
  def __call__(self, json_request: JSON_TYPE) -> JSON_TYPE:
    result = requests.post(self._endpoint, json=json_request)
    return json.loads(result.content)
```

and then

```py
class QuokkaSampler(cirq.Sampler):
  def __init__(self, post: Optional[Callable[[JSON_TYPE], JSON_TYPE]] = None):
    self._post = post or Quokka()
```

Now there's a clean way to instantiate sampler with different post methods:

```py
s = QuokkaSampler()  # production instance
s = QuokkaSampler(post=_mock_post)  # unittest post function
s = QuokkaSampler(post=Quokka("staging"))  # staging quokka
s = QuokkaSampler(post=Quokka("foo"))  # foo quokka
```

in fact, you're already doing something very similar in test, so this refactoring makes even more sense :)

Reply:
Done.  Kept name as a parameter though since QuokkaSample("foo") seemed cleaner.